docs/proposals/blocks-storage-sharding.md

```diff
@@ -0,0 +1,77 @@
+---
+title: "Blocks storage sharding"
+linkTitle: "Blocks storage sharding"
+weight: 1
+slug: blocks-storage-sharding
+---
+
+- Author: [Marco Pracucci](https://github.com/pracucci)
+- Date: March 2020
+- Status: proposal
+
+## Problem
+
+In Cortex, when using the experimental blocks storage, each querier internally runs the Thanos [`BucketStore`](https://github.com/thanos-io/thanos/blob/master/pkg/store/bucket.go). This means that each querier has a full view over all blocks in the long-term storage and all blocks index headers are loaded in each querier memory. The querier memory usage linearly increase with number and size of all blocks in the storage, imposing a scalability limit to the blocks storage.
+
+In this proposal we want to solve this. In particular, we want to:
+
+1. Shard blocks (index headers) across a pool of nodes
+2. Do not compromise HA on the read path (if a node fails, queries should continue to work)
+3. Do not compromise correctness (either the query result is correct or it fails)
+
+## Proposed solution
+
+The idea is to introduce a new Cortex service - `store-gateway` - internally running the Thanos [`BucketStore`](https://github.com/thanos-io/thanos/blob/master/pkg/store/bucket.go). At query time, a querier will run a query fetching the matching series both from ingesters and the subset of gateways holding the related blocks (based on the query time range). Blocks are replicated across the gateways in order to guarantee query results consistency and HA even in the case of a gateway instance failure.
+
+### Ring-based sharding and replication
```

Comment:
Are store-gateways going to transfer blocks within them when joining / leaving the ring? Would it make sense to do so?

Reply:
Looks an unnecessary complexity. From a performances perspective, I don't expect any significant difference transferring them between store-gateways instead of re-downloading them from the storage.